src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs

```diff
@@ -633,41 +652,108 @@ private void WriteFile()
+                            // inflation and deflation.
+                            if ((entry.Changes & (ChangeState.DynamicLengthMetadata | ChangeState.StoredData)) != 0)
+                            {
+                                completeRewriteStartingOffset = Math.Min(completeRewriteStartingOffset, entry.OffsetOfLocalHeader);
+                            }
+                            if (entry.OffsetOfLocalHeader >= completeRewriteStartingOffset)
+                            {
+                                entry.LoadLocalHeaderExtraFieldAndCompressedBytesIfNeeded();
+                            }
+
+                            entriesToWrite.Add(entry);
+                        }
+                    }
+                }
+
+                // If the offset of entries to write from is still at long.MaxValue, then we know that nothing has been deleted,
+                // nothing has been modified - so we just want to move to the end of all remaining files in the archive.
+                if (startingOffset == long.MaxValue)
+                {
+                    startingOffset = nextFileOffset;
                 }
-                foreach (ZipArchiveEntry entry in markedForDelete)
-                    entry.Delete();
 
-                _archiveStream.Seek(0, SeekOrigin.Begin);
-                _archiveStream.SetLength(0);
+                _archiveStream.Seek(startingOffset, SeekOrigin.Begin);
             }
 
-            foreach (ZipArchiveEntry entry in _entries)
+            foreach (ZipArchiveEntry entry in entriesToWrite)
             {
-                entry.WriteAndFinishLocalEntry();
+                // We don't always need to write the local header entry, ZipArchiveEntry is usually able to work out when it doesn't need to.
+                // We want to force this header entry to be written (even for completely untouched entries) if the entry comes after one
+                // which had a pending dynamically-sized write.
+                bool forceWriteLocalEntry = !entry.OriginallyInArchive || (entry.OriginallyInArchive && entry.OffsetOfLocalHeader >= completeRewriteStartingOffset);
+
+                entry.WriteAndFinishLocalEntry(forceWriteLocalEntry);
             }
 
             long startOfCentralDirectory = _archiveStream.Position;
+            // If there are no entries in the archive, we still want to create the archive epilogue.
+            bool archiveEpilogueRequiresUpdate = _entries.Count == 0;
 
             foreach (ZipArchiveEntry entry in _entries)
             {
-                entry.WriteCentralDirectoryFileHeader();
+                // The central directory needs to be rewritten if its position has moved, if there's a new entry in the archive, or if the entry might be different.
+                bool centralDirectoryEntryRequiresUpdate = startOfCentralDirectory != _centralDirectoryStart
```

Comment:
Thanks - I agree it's confusing. I've renamed `startOfCentralDirectory` to `plannedCentralDirectoryPosition`. The instance variable `_centralDirectoryStart` contains the current position of the central directory.

Reply:
I'd prefer `plannedStartOfCentralDirectory`. Or maybe `newStartOfCentralDirectory` would be better: shorter and still indicates it will be the new value. Would you agree?

Reply:
I'm ok either way. I see you already pushed the commit with `plannedCentralDirectoryPosition `, which is good too.